src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/OnSerializeTests.cs

```diff
@@ -0,0 +1,430 @@
+            {
+                Assert.Equal(1, _onDeserializingCount);
+                _onDeserializedCount++;
+                Assert.Equal(1, MyInt);
+                MyInt++;
+            }
+        }
+
+        [Fact]
+        public static void Test_MyCyclicClass()
+        {
+            const string Json = "{\"$id\":\"1\",\"MyInt\":1,\"Cycle\":{\"$ref\":\"1\"}}";
+
+            MyCyclicClass obj = new();
+            obj.MyInt = 1;
+            obj.Cycle = obj;
+
+            JsonSerializerOptions options = new();
+            options.ReferenceHandler = ReferenceHandler.Preserve;
+
+            string json = JsonSerializer.Serialize(obj, options);
+            Assert.Contains("\"MyInt\":2", json);
+            Assert.Equal(1, obj._onSerializingCount);
+            Assert.Equal(1, obj._onSerializedCount);
+            Assert.Equal(0, obj._onDeserializingCount);
+            Assert.Equal(0, obj._onDeserializedCount);
+
+            obj = JsonSerializer.Deserialize<MyCyclicClass>(Json, options);
+            Assert.Equal(2, obj.MyInt);
+            Assert.Equal(obj, obj.Cycle);
+            Assert.Equal(0, obj._onSerializingCount);
+            Assert.Equal(0, obj._onSerializedCount);
+            Assert.Equal(1, obj._onDeserializingCount);
+            Assert.Equal(1, obj._onDeserializedCount);
+        }
+
+        private class MyCollection : List<int>,
+            IJsonOnDeserializing,
+            IJsonOnDeserialized,
+            IJsonOnSerializing,
+            IJsonOnSerialized
+        {
+            public void OnDeserialized() => Assert.True(false, "Not expected");
+            public void OnDeserializing() => Assert.True(false, "Not expected");
+            public void OnSerialized() => Assert.True(false, "Not expected");
+            public void OnSerializing() => Assert.True(false, "Not expected");
+        }
+
+        [JsonConverter(converterType: typeof(MyValueConverter))]
+        private class MyValue :
+            IJsonOnDeserializing,
+            IJsonOnDeserialized,
+            IJsonOnSerializing,
+            IJsonOnSerialized
+        {
+            public void OnDeserialized() => Assert.True(false, "Not expected");
+            public void OnDeserializing() => Assert.True(false, "Not expected");
+            public void OnSerialized() => Assert.True(false, "Not expected");
+            public void OnSerializing() => Assert.True(false, "Not expected");
+        }
+
+        private class MyValueConverter : JsonConverter<MyValue>
+        {
+            public override MyValue? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                return new MyValue();
+            }
+
+            public override void Write(Utf8JsonWriter writer, MyValue value, JsonSerializerOptions options)
+            {
+                writer.WriteStringValue("dummy");
+            }
+        }
+
+        [Fact]
+        public static void NonPocosIgnored()
```

Comment:
What prevents us from supporting callbacks in collection types?

Reply:
Nothing (same for value types) although there are very limited scenarios for both since custom converters would normally be used which implicitly have support for before\after.

Also, a richer interaction model would be better for collections - in order to provide support for "Add()" as well as holding onto any temporary collections needed for things like Arrays, Stacks and immutable collections.